edge-util/src/Microsoft.Azure.Devices.Edge.Storage/DbStoreProviderWithBackupRestore.cs

```diff
@@ -0,0 +1,351 @@
+                this.CleanupAllBackups(this.backupPath);
+            }
+        }
+
+        public override IDbStore GetDbStore(string partitionName)
+        {
+            this.dbStores.Add(partitionName);
+            return base.GetDbStore(partitionName);
+        }
+
+        public override IDbStore GetDbStore()
+        {
+            this.dbStores.Add(DefaultStoreBackupName);
+            return base.GetDbStore();
+        }
+
+        public override void RemoveDbStore(string partitionName)
+        {
+            this.dbStores.Remove(partitionName);
+            base.RemoveDbStore(partitionName);
+        }
+
+        public override void RemoveDbStore()
+        {
+            this.dbStores.Remove(DefaultStoreBackupName);
+            base.RemoveDbStore();
+        }
+
+        public async override Task CloseAsync()
+        {
+            await this.BackupAsync();
+            await base.CloseAsync();
+        }
+
+        private async Task BackupAsync()
+        {
+            this.events.StartingBackup();
+            Guid backupId = Guid.NewGuid();
+            string dbBackupDirectory = Path.Combine(this.backupPath, backupId.ToString());
+
+            BackupMetadata newBackupMetadata = new BackupMetadata(backupId, this.backupFormat, DateTime.UtcNow, this.dbStores.ToList());
+            BackupMetadataList backupMetadataList = new BackupMetadataList(new List<BackupMetadata> { newBackupMetadata });
+            try
+            {
+                Directory.CreateDirectory(dbBackupDirectory);
+
+                // Backup other stores.
+                foreach (string store in this.dbStores)
+                {
+                    IDbStore dbStore = this.dbStoreProvider.GetDbStore(store);
+                    await this.dbStoreBackupRestore.BackupAsync(store, dbStore, dbBackupDirectory);
+                }
+
+                using (StreamWriter file = File.CreateText(Path.Combine(this.backupPath, BackupMetadataFileName)))
+                {
+                    JsonSerializer serializer = new JsonSerializer();
+                    serializer.Serialize(file, backupMetadataList);
+                }
+
+                this.events.BackupComplete();
+
+                // Clean any old backups.
+                this.CleanupUnknownBackups(this.backupPath, backupMetadataList);
+            }
+            catch (Exception exception)
+            {
+                this.events.BackupFailure($"The backup operation failed with error ${exception}.");
+
+                // Clean up any artifacts of the attempted backup.
+                this.CleanupKnownBackups(this.backupPath, backupMetadataList);
+            }
+        }
+
+        private void CleanupAllBackups(string backupPath)
+        {
+            DirectoryInfo backupDirInfo = new DirectoryInfo(backupPath);
+            foreach (FileInfo file in backupDirInfo.GetFiles())
+            {
+                file.Delete();
+            }
```

Comment:
maybe catch all exceptions and continue the loop? I can imagine somehow a restricted file or directory gets there accidentally and then the backup files never will be deleted because this loop exist with an exception?

Reply:
Yep, sounds good. We can log when we fail to delete some of the files in there. I'll make that change.